### DISCUSSION TITLE
Getting decompilation to display strings assigned via SIMD instruction

### DISCUSSION THREAD

**Discussion by leechristensen** [NationalSecurityAgency/ghidra#6925] [Q&A] [answered]
How does one get the decomplication to to display strings when SIMD instructions like MOVUPS/MOVAPS are used to assign the string? For example, in the screenshot below, we can see in the assembly that the data is defined as a unicode string but the decompilation doesn't display the string:

![image](https://github.com/user-attachments/assets/627ccfda-f2a3-4713-894c-1e4be2045a03)

Relevant as well: https://reverseengineering.stackexchange.com/questions/24974/how-to-improve-ghidra-decompilation-of-sse-instructions

**Comment by Wall-AF** [comment]
Try marking the data as `constant`.  Right click on the data (in the `Listing:`Panel), go into the `Data` submenu, open either the `Settings...` (for a one off) or `Default Settings...` dialog, change the `Mutability` to `Constant`.
![image](https://github.com/user-attachments/assets/414aad9e-8f62-4f0a-afd5-fbc3c864186a)

**Reply by leechristensen** [reply]
Yeah, I had tried that as well with no luck:
![image](https://github.com/user-attachments/assets/618dc445-efd3-4b74-b129-87fd153372f6)

**Reply by leechristensen** [reply]
[Here's the target file](https://msdl.microsoft.com/download/symbols/ncryptprov.dll/529E02C06a000/ncryptprov.dll) in case anyone wants to play along. See the `GetUserDirectory` function

**Comment by emteere** [comment, marked as answer]
In the latest Ghidra 11.2, which should be released in a few days, and the main branch, the decompiler now recovers stack strings and there is initial support for optimized heap strings.  The implementation may require more work for strings built using SIMD registers, however it might work for your case.

**Reply by leechristensen** [reply]
Thanks, I'll give it a try when it's released and report back!

**Reply by leechristensen** [reply]
The default analysis still results in a similar issue due to it not recognizing `puVar1`'s type (it's an undefined pointer):
![image](https://github.com/user-attachments/assets/1a4ce4e2-df87-47cf-acf1-3dcfcf74a39c)

Switching `puVar1` to a `LPCWSTR` fixes it up though!
![image](https://github.com/user-attachments/assets/85e82146-9e9d-4b80-928e-e65d9f4ccea3)

FWIW, both BinaryNinja/Ida pickup the string type default. Should this still be filed as an issue, or something separate?

**Comment by emteere** [comment]
I think detecting the type of the variable from the string already created in memory would be a worthwhile change.  I can't speak to how easy that change would be.  I've created an issue from the discussion #6985.  Feel free to augment the issue.

I've also put in an internal ticket to consider using the fact that a string has been defined in memory that is being loaded into the SIMD register.
